mailbox/event/event-memory/src/main/java/org/apache/james/mailbox/events/delivery/InVmEventDelivery.java

```diff
@@ -72,19 +73,29 @@ public ExecutionStages deliver(Collection<MailboxListener> mailboxListeners, Eve
 
     private Mono<Void> doDeliver(Collection<MailboxListener> mailboxListeners, Event event) {
         return Flux.fromIterable(mailboxListeners)
-            .flatMap(mailboxListener -> Mono.fromRunnable(() -> doDeliverToListener(mailboxListener, event)))
+            .flatMap(mailboxListener -> deliveryWithRetries(event, mailboxListener))
             .then()
             .subscribeOn(Schedulers.elastic());
     }
 
+    private Mono<Object> deliveryWithRetries(Event event, MailboxListener mailboxListener) {
+        return Mono.fromRunnable(() -> doDeliverToListener(mailboxListener, event))
+            .retry(MAX_RETRIES)
+            .doOnError(throwable -> LOGGER.error("listener {} exceeded maximum retry({}) to handle event {}",
+                mailboxListener.getClass().getCanonicalName(),
+                MAX_RETRIES,
+                event.getClass().getCanonicalName(),
+                throwable));
+    }
+
     private void doDeliverToListener(MailboxListener mailboxListener, Event event) {
         TimeMetric timer = metricFactory.timer("mailbox-listener-" + mailboxListener.getClass().getSimpleName());
         try {
             mailboxListener.event(event);
-        } catch (Throwable throwable) {
+        } catch (Exception ex) {
             LOGGER.error("Error while processing listener {} for {}",
-                mailboxListener.getClass().getCanonicalName(), event.getClass().getCanonicalName(),
-                throwable);
+                mailboxListener.getClass().getCanonicalName(), event.getClass().getCanonicalName(), ex);
+            throw ex;
```

Comment:
Log then trhow seems like a bad pattern with me. Why not rely on logger line 84 in deliveryWithRetries ?